File: python/automarketing/media/hosts.py
"""
Public hosting — Instagram/Facebook ne file aapva mate.

⚠️ AA AAKHA SYSTEM NU SAUTHI MOTU "GOTCHA":
   Meta na server AAPNI file DOWNLOAD kare che. Etle
   `http://localhost:8000/...` KYAREY nahi chale — public https URL joiye j.

Etle ek chain rakhi che. Uper thi niche — je chale e:

  1. base-url   — tamaru potanu domain / ngrok tunnel (sauthi saru, free)
  2. catbox     — KOI KEY NAHI, image + video, kayami rahe che
  3. cloudinary — free 25GB CDN (key joiye pan free che)
  4. tmpfiles   — koi key nahi, pan file fakt 1 kalak rahe che

Tamare kai key na joiti hoy to pan chale — catbox key vagar j chale che.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Optional

from ..config import settings
from ..errors import FatalError
from ..pipeline.chain import Candidate, ChainResult, run_chain
from ..pipeline.http import get_client


@dataclass
class UploadedFile:
    url: str
    host: str
    #: Aa URL kyare khatam thashe (khabar hoy to).
    expires_at: Optional[datetime] = None


@dataclass
class UploadInput:
    data: bytes
    filename: str
    mime_type: str
    #: "image" | "video" | "audio"
    kind: str = "image"


# ------------------------------------------------------------------ #
#  Catbox — koi key nahi (200MB sudhi)
# ------------------------------------------------------------------ #


async def _upload_catbox(item: UploadInput) -> UploadedFile:
    client = get_client()
    response = await client.post(
        "https://catbox.moe/user/api.php",
        data={"reqtype": "fileupload"},
        files={"fileToUpload": (item.filename, item.data, item.mime_type)},
        timeout=300.0,
    )

    text = (response.text or "").strip()
    if response.status_code >= 400 or not text.startswith("https://"):
        raise FatalError(f"Catbox: {text[:200] or response.status_code}")

    return UploadedFile(url=text, host="catbox")


# ------------------------------------------------------------------ #
#  tmpfiles — koi key nahi, 1 kalak
# ------------------------------------------------------------------ #


async def _upload_tmpfiles(item: UploadInput) -> UploadedFile:
    client = get_client()
    response = await client.post(
        "https://tmpfiles.org/api/v1/upload",
        files={"file": (item.filename, item.data, item.mime_type)},
        timeout=300.0,
    )
    if response.status_code >= 400:
        raise FatalError(f"tmpfiles: HTTP {response.status_code}")

    page = ((response.json() or {}).get("data") or {}).get("url")
    if not page:
        raise FatalError("tmpfiles: URL na madyu")

    # tmpfiles page nu URL aape che — direct download mate `/dl/` joiye.
    direct = page.replace("tmpfiles.org/", "tmpfiles.org/dl/", 1)
    return UploadedFile(
        url=direct,
        host="tmpfiles",
        expires_at=datetime.now(timezone.utc) + timedelta(minutes=55),
    )


# ------------------------------------------------------------------ #
#  Cloudinary — free 25GB CDN
# ------------------------------------------------------------------ #


def _cloudinary_ready() -> bool:
    return bool(settings.cloudinary_cloud and settings.cloudinary_preset)


async def _upload_cloudinary(item: UploadInput) -> UploadedFile:
    resource = "video" if item.kind in ("video", "audio") else "image"
    url = f"https://api.cloudinary.com/v1_1/{settings.cloudinary_cloud}/{resource}/upload"

    client = get_client()
    response = await client.post(
        url,
        data={"upload_preset": settings.cloudinary_preset},
        files={"file": (item.filename, item.data, item.mime_type)},
        timeout=300.0,
    )

    payload = {}
    try:
        payload = response.json() or {}
    except Exception:  # noqa: BLE001
        pass

    secure_url = payload.get("secure_url")
    if response.status_code >= 400 or not secure_url:
        message = ((payload.get("error") or {}).get("message")) or response.text[:200]
        raise FatalError(f"Cloudinary: {message}")

    return UploadedFile(url=secure_url, host="cloudinary")


# ------------------------------------------------------------------ #
#  Chain
# ------------------------------------------------------------------ #


async def upload_public(
    item: UploadInput,
    *,
    prefer: Optional[str] = None,
) -> ChainResult[UploadedFile]:
    """File ne public URL par mukho. Je host chale e vaparashe."""
    anon = settings.allow_anon_hosts

    candidates = [
        Candidate[UploadedFile](
            name="cloudinary",
            label="Cloudinary (free 25GB CDN)",
            configured=_cloudinary_ready,
            run=lambda: _upload_cloudinary(item),
            timeout=300.0,
        ),
        Candidate[UploadedFile](
            name="catbox",
            label="Catbox (key vagar)",
            configured=lambda: anon,
            run=lambda: _upload_catbox(item),
            timeout=300.0,
        ),
        Candidate[UploadedFile](
            name="tmpfiles",
            label="tmpfiles.org (key vagar, 1 kalak)",
            configured=lambda: anon,
            run=lambda: _upload_tmpfiles(item),
            timeout=300.0,
        ),
    ]

    return await run_chain(
        candidates,
        label="Public media hosting",
        prefer=(prefer or settings.preferred_host),
        retries=1,
        backoff=1.5,
    )


def host_status() -> list[dict]:
    """Setup page mate."""
    anon = settings.allow_anon_hosts
    base = settings.public_media_base_url
    return [
        {
            "key": "base-url",
            "label": "Potanu domain / ngrok tunnel",
            "free": True,
            "configured": bool(base.startswith("https://")),
            "recommended": True,
            "note": (
                "SAUTHI SARU ane sav free. `ngrok http 8000` chalavo ane e https "
                "URL .env ma PUBLIC_MEDIA_BASE_URL ma nakho. Pachi koi upload "
                "karvani jarur j nathi."
            ),
        },
        {
            "key": "catbox",
            "label": "Catbox",
            "free": True,
            "configured": anon,
            "recommended": True,
            "note": "Koi key nahi — turant chale che. Image + video, kayami rahe che.",
        },
        {
            "key": "cloudinary",
            "label": "Cloudinary",
            "free": True,
            "configured": _cloudinary_ready(),
            "recommended": False,
            "note": (
                "Free 25GB CDN. cloudinary.com par signup → Settings → Upload → "
                "Add upload preset → Signing Mode: Unsigned."
            ),
        },
        {
            "key": "tmpfiles",
            "label": "tmpfiles.org",
            "free": True,
            "configured": anon,
            "recommended": False,
            "note": "Koi key nahi, pan file fakt 1 kalak rahe che. Chhelli aasha.",
        },
    ]


def cache_key(data: bytes) -> str:
    """Ek j file be var upload na thay etle."""
    return hashlib.sha256(data).hexdigest()[:32]

File: src/lib/media/hosts.ts
/**
 * Public hosting — Instagram/Facebook ne file aapva mate.
 *
 * ⚠️ Aa aakha system nu sauthi motu "gotcha": Meta na server aapni file
 * download kare che. Etle `http://localhost:3000/...` KYAREY nahi chale —
 * public https URL joiye j.
 *
 * Etle ahiya ek chain rakhi che. Uper thi niche — je configure hoy e chale:
 *
 *   1. cloudinary — best (free 25GB, saacho CDN, video support, kayami URL)
 *   2. imgbb      — free key, fakt image
 *   3. catbox     — koi key nahi, image + video, anonymous
 *   4. tmpfiles   — koi key nahi, 1 kalak ni file (chhelli aasha)
 *   5. base-url   — tamaru potanu domain / ngrok tunnel
 *
 * Ek pan na chale to samjay evo error aave che — chup-chaap fail nahi.
 */

import { apiFetch, runChain, FatalError, type ChainResult } from "@/lib/pipeline/chain";

export type UploadedFile = {
  url: string;
  host: string;
  /** Aa URL kyare khatam thashe (khabar hoy to). */
  expiresAt?: Date;
};

export type UploadInput = {
  data: Buffer;
  filename: string;
  mimeType: string;
  kind: "image" | "video" | "audio" | "other";
};

/* ------------------------------------------------------------------ *
 *  Cloudinary — recommended
 * ------------------------------------------------------------------ */

function cloudinaryConfig() {
  return {
    cloud: process.env.CLOUDINARY_CLOUD_NAME || "",
    preset: process.env.CLOUDINARY_UPLOAD_PRESET || "",
    apiKey: process.env.CLOUDINARY_API_KEY || "",
    apiSecret: process.env.CLOUDINARY_API_SECRET || "",
  };
}

function cloudinaryConfigured(): boolean {
  const c = cloudinaryConfig();
  return Boolean(c.cloud && (c.preset || (c.apiKey && c.apiSecret)));
}

async function uploadToCloudinary(
  input: UploadInput,
  signal: AbortSignal,
): Promise<UploadedFile> {
  const c = cloudinaryConfig();
  const resourceType = input.kind === "video" || input.kind === "audio" ? "video" : "image";
  const url = `https://api.cloudinary.com/v1_1/${c.cloud}/${resourceType}/upload`;

  const form = new FormData();
  form.append("file", new Blob([new Uint8Array(input.data)], { type: input.mimeType }), input.filename);

  if (c.preset) {
    // Unsigned upload — sauthi saral, fakt preset joiye.
    form.append("upload_preset", c.preset);
  } else {
    // Signed upload — api key + secret thi signature banaviye.
    const timestamp = Math.floor(Date.now() / 1000);
    const { createHash } = await import("node:crypto");
    const signature = createHash("sha1")
      .update(`timestamp=${timestamp}${c.apiSecret}`)
      .digest("hex");
    form.append("api_key", c.apiKey);
    form.append("timestamp", String(timestamp));
    form.append("signature", signature);
  }

  const json = await apiFetch<{ secure_url?: string; error?: { message: string } }>(url, {
    method: "POST",
    body: form,
    signal,
  });

  if (!json.secure_url) {
    throw new FatalError(`Cloudinary: ${json.error?.message ?? "URL na madyu"}`);
  }
  return { url: json.secure_url, host: "cloudinary" };
}

/* ------------------------------------------------------------------ *
 *  ImgBB — free key, fakt image
 * ------------------------------------------------------------------ */

async function uploadToImgbb(
  input: UploadInput,
  signal: AbortSignal,
): Promise<UploadedFile> {
  const key = process.env.IMGBB_API_KEY || "";
  const form = new FormData();
  form.append("image", input.data.toString("base64"));
  form.append("name", input.filename.replace(/\.[^.]+$/, ""));

  const json = await apiFetch<{
    data?: { url?: string; display_url?: string };
    error?: { message: string };
  }>(`https://api.imgbb.com/1/upload?key=${encodeURIComponent(key)}`, {
    method: "POST",
    body: form,
    signal,
  });

  const url = json.data?.display_url ?? json.data?.url;
  if (!url) throw new FatalError(`ImgBB: ${json.error?.message ?? "URL na madyu"}`);
  return { url, host: "imgbb" };
}

/* ------------------------------------------------------------------ *
 *  Catbox — koi key nahi, image + video (200MB sudhi)
 * ------------------------------------------------------------------ */

async function uploadToCatbox(
  input: UploadInput,
  signal: AbortSignal,
): Promise<UploadedFile> {
  const form = new FormData();
  form.append("reqtype", "fileupload");
  form.append(
    "fileToUpload",
    new Blob([new Uint8Array(input.data)], { type: input.mimeType }),
    input.filename,
  );

  const text = await apiFetch<string>("https://catbox.moe/user/api.php", {
    method: "POST",
    body: form,
    signal,
    expect: "text",
  });

  const url = text.trim();
  if (!url.startsWith("https://")) {
    throw new FatalError(`Catbox: ${url.slice(0, 200)}`);
  }
  return { url, host: "catbox" };
}

/* ------------------------------------------------------------------ *
 *  tmpfiles.org — koi key nahi, 60 minute sudhi rahe
 * ------------------------------------------------------------------ */

async function uploadToTmpfiles(
  input: UploadInput,
  signal: AbortSignal,
): Promise<UploadedFile> {
  const form = new FormData();
  form.append(
    "file",
    new Blob([new Uint8Array(input.data)], { type: input.mimeType }),
    input.filename,
  );

  const json = await apiFetch<{ data?: { url?: string }; status?: string }>(
    "https://tmpfiles.org/api/v1/upload",
    { method: "POST", body: form, signal },
  );

  const page = json.data?.url;
  if (!page) throw new FatalError("tmpfiles: URL na madyu");

  // tmpfiles page URL aape che — direct download mate `/dl/` umervu pade.
  const direct = page.replace("tmpfiles.org/", "tmpfiles.org/dl/");
  return {
    url: direct,
    host: "tmpfiles",
    expiresAt: new Date(Date.now() + 55 * 60_000),
  };
}

/* ------------------------------------------------------------------ *
 *  Chain
 * ------------------------------------------------------------------ */

export type PublicHostKey =
  | "cloudinary"
  | "imgbb"
  | "catbox"
  | "tmpfiles"
  | "base-url";

/**
 * File ne public URL par mukho. Configure thayela hosts kram ma try thay che.
 *
 * `preferHost` aapo to e pehla try thay (setup page ni pasandgi).
 */
export async function uploadPublic(
  input: UploadInput,
  preferHost?: string,
): Promise<ChainResult<UploadedFile>> {
  const isImage = input.kind === "image";

  return runChain<UploadedFile>(
    [
      {
        name: "cloudinary",
        label: "Cloudinary (free 25GB CDN)",
        free: true,
        configured: cloudinaryConfigured,
        run: (signal) => uploadToCloudinary(input, signal),
        timeoutMs: 180_000,
      },
      {
        name: "imgbb",
        label: "ImgBB (free, fakt image)",
        free: true,
        configured: () => Boolean(process.env.IMGBB_API_KEY) && isImage,
        run: (signal) => uploadToImgbb(input, signal),
        timeoutMs: 120_000,
      },
      {
        name: "catbox",
        label: "Catbox (key vagar)",
        free: true,
        configured: () => process.env.MEDIA_ALLOW_ANON_HOSTS !== "false",
        run: (signal) => uploadToCatbox(input, signal),
        timeoutMs: 180_000,
      },
      {
        name: "tmpfiles",
        label: "tmpfiles.org (key vagar, 1 kalak)",
        free: true,
        configured: () => process.env.MEDIA_ALLOW_ANON_HOSTS !== "false",
        run: (signal) => uploadToTmpfiles(input, signal),
        timeoutMs: 180_000,
      },
    ],
    {
      label: "Public media hosting",
      prefer: preferHost ?? process.env.MEDIA_HOST,
      retries: 1,
      backoffMs: 1200,
    },
  );
}

/** Setup page mate — kayo host taiyar che. */
export function hostStatus() {
  const anon = process.env.MEDIA_ALLOW_ANON_HOSTS !== "false";
  return [
    {
      key: "cloudinary" as const,
      label: "Cloudinary",
      free: true,
      configured: cloudinaryConfigured(),
      recommended: true,
      note: "Sauthi saaru — free 25GB, video support, kayami URL. cloudinary.com par signup → Settings → Upload → unsigned preset banavo.",
    },
    {
      key: "imgbb" as const,
      label: "ImgBB",
      free: true,
      configured: Boolean(process.env.IMGBB_API_KEY),
      recommended: false,
      note: "Fakt image. api.imgbb.com par thi free key.",
    },
    {
      key: "catbox" as const,
      label: "Catbox",
      free: true,
      configured: anon,
      recommended: false,
      note: "Koi key nahi — turant chale. Image + video.",
    },
    {
      key: "tmpfiles" as const,
      label: "tmpfiles.org",
      free: true,
      configured: anon,
      recommended: false,
      note: "Koi key nahi, pan file fakt 1 kalak rahe che. Chhelli aasha.",
    },
    {
      key: "base-url" as const,
      label: "Potanu domain / tunnel",
      free: true,
      configured: Boolean(process.env.PUBLIC_MEDIA_BASE_URL),
      recommended: false,
      note: "PUBLIC_MEDIA_BASE_URL set karo (dakhla tarike ngrok ke tamaru domain) to /api/media sidhu vaparashe.",
    },
  ];
}
